apps/emqx_rule_engine/test/emqx_rulesql_SUITE.erl

```diff
@@ -523,9 +538,122 @@ t_sqlparse_event_client_connected_02(_Config) ->
+    {ok, _, _} = emqtt:subscribe(Client, RepubT, 0),
+    ct:sleep(200),
+    {ok, Client1} = emqtt:start_link([{clientid, <<"emqx">>}, {username, <<"emqx">>}]),
+    {ok, _} = emqtt:connect(Client1),
+    emqtt:disconnect(Client1),
+
+    receive {publish, #{topic := T, payload := Payload}} ->
+        ?assertEqual(RepubT, T),
+        ?assertMatch(#{<<"reason">> := <<"normal">>}, emqx_json:decode(Payload, [return_maps]))
+    after 1000 ->
+        ct:fail(wait_for_repub_disconnected_normal)
+    end,
+    emqtt:stop(Client),
+
+    emqx_rule_registry:remove_rule(TopicRule).
+
+t_sqlparse_event_client_disconnected_kicked(_Config) ->
+    process_flag(trap_exit, true),
+    ok = emqx_rule_engine:load_providers(),
+    Sql = "select * "
+          "from \"$events/client_disconnected\" ",
+    RepubT = <<"repub/to/disconnected/kicked">>,
+
+    TopicRule = create_simple_repub_rule(RepubT, Sql, <<>>),
+
+    {ok, Client} = emqtt:start_link([{clientid, <<"get_repub_client">>}, {username, <<"emqx0">>}]),
+    {ok, _} = emqtt:connect(Client),
+    {ok, _, _} = emqtt:subscribe(Client, RepubT, 0),
+    ct:sleep(200),
+    {ok, Client1} = emqtt:start_link([{clientid, <<"emqx">>}, {username, <<"emqx">>}]),
+    {ok, _} = emqtt:connect(Client1),
+    emqx_cm:kick_session(<<"emqx">>),
+
+    receive {publish, #{topic := T, payload := Payload}} ->
+        ?assertEqual(RepubT, T),
+        ?assertMatch(#{<<"reason">> := <<"kicked">>}, emqx_json:decode(Payload, [return_maps]))
+    after 1000 ->
+        ct:fail(wait_for_repub_disconnected_kicked)
+    end,
+    emqtt:stop(Client),
+
+    process_flag(trap_exit, false).
+
+t_sqlparse_event_client_disconnected_discarded(_Config) ->
+    process_flag(trap_exit, true),
+    ok = emqx_rule_engine:load_providers(),
+    Sql = "select * "
+          "from \"$events/client_disconnected\" ",
+    RepubT = <<"repub/to/disconnected/discarded">>,
+
+    TopicRule = create_simple_repub_rule(RepubT, Sql, <<>>),
+
+    {ok, Client} = emqtt:start_link([{clientid, <<"get_repub_client">>}, {username, <<"emqx0">>}]),
+    {ok, _} = emqtt:connect(Client),
+    {ok, _, _} = emqtt:subscribe(Client, RepubT, 0),
+    ct:sleep(200),
+
+    {ok, Client1} = emqtt:start_link([{clientid, <<"emqx">>}, {username, <<"emqx">>}]),
+    {ok, _} = emqtt:connect(Client1),
+
+    {ok, Client2} = emqtt:start_link([{clientid, <<"emqx">>}, {username, <<"emqx">>}, {clean_start, true}]),
+    {ok, _} = emqtt:connect(Client2),
+
+    receive {publish, #{topic := T, payload := Payload}} ->
+        ?assertEqual(RepubT, T),
+        ?assertMatch(#{<<"reason">> := <<"discarded">>}, emqx_json:decode(Payload, [return_maps]))
+    after 1000 ->
+        ct:fail(wait_for_repub_disconnected_discarded)
+    end,
+    emqtt:stop(Client),
+    emqtt:stop(Client2),
+
+    emqx_rule_registry:remove_rule(TopicRule),
+    process_flag(trap_exit, false).
+
+t_sqlparse_event_client_disconnected_takeovered(_Config) ->
+    process_flag(trap_exit, true),
```

Comment:
do not trap exit, instead, explicitly unlink the pids which may cause test to exit.

even better: make sure the pid is down:
```
kill_process(Pid) ->
    kill_process(Pid, fun(_) -> erlang:exit(Pid, kill) end).

kill_process(Pid, WithFun) ->
    _ = unlink(Pid),
    _ = monitor(process, Pid),
    _ = WithFun(Pid),
    receive
        {'DOWN', _, process, Pid, _} ->
            ok
    after 10_000 ->
              error(timeout)
    end.
```